### PR TITLE
Authentication Screen - Backend authentication - Login screen with validator

### PR DESCRIPTION
**Pull request by AnhToof**
https://jobchat.vn/job/-LPOJhtoknDaKAnQF91f/authentication-screen---backend-authentication-LREkZquAocJHV6iHif-

### CODE REVIEW

File: app/src/main/java/com/lobesoftware/toof/firebase_chat_001/extension/ActivityExtension.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+fun AppCompatActivity.replaceFragment(
+    @IdRes containerId: Int, fragment: Fragment,
+    addToBackStack: Boolean = false,
+    tag: String = fragment::class.java.simpleName
+) {
+    val fragmentManager = supportFragmentManager
+    val transaction = fragmentManager.beginTransaction()
+    if (addToBackStack) transaction.addToBackStack(tag)
```
Comment (onstonboy) [inline app/src/main/java/com/lobesoftware/toof/firebase_chat_001/extension/ActivityExtension.kt]: should put codes in block if (){}

File: app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginFragment.kt [hunk 2]
```diff
@@ -1,18 +1,73 @@
...
+import com.lobesoftware.toof.firebase_chat_001.utils.validator.Validator
+import kotlinx.android.synthetic.main.fragment_login.*
+import javax.inject.Inject
 
-class LoginFragment : Fragment() {
+class LoginFragment : Fragment(), LoginContract.View {
+
+    private lateinit var mPresenter: LoginPresenter
+
+    @Inject
```
Comment (onstonboy) [inline app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginFragment.kt]: should be put @Inject ... in top define (convention)

File: app/src/main/java/com/lobesoftware/toof/firebase_chat_001/extension/ActivityExtension.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package com.lobesoftware.toof.firebase_chat_001.extension
+
+import android.content.Intent
+import android.support.annotation.IdRes
+import android.support.v4.app.Fragment
+import android.support.v7.app.AppCompatActivity
+
+fun AppCompatActivity.startActivity(intent: Intent, flags: Int? = null) {
+    flags?.let { intent.flags = flags }
```
Comment (truyenz) [inline app/src/main/java/com/lobesoftware/toof/firebase_chat_001/extension/ActivityExtension.kt]: flags?.let { intent.flags = it } or flags?.let { name_variable -> intent.flags = name_variable }

File: app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginFragment.kt [hunk 4]
```diff
@@ -1,18 +1,73 @@
...
         savedInstanceState: Bundle?
     ): View? {
-        return inflater.inflate(R.layout.fragment_login, container, false)
+        val view = inflater.inflate(R.layout.fragment_login, container, false)
+
+        mPresenter = LoginPresenter()
+        mPresenter.setValidator(mValidator)
+        mPresenter.setView(this)
+
+        view.findViewById<Button>(R.id.button_sign_in).setOnClickListener {
```
Comment (onstonboy) [inline app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginFragment.kt]: should be import kotlinx.android.synthetic..... LoginFragment.view.* to use these view. not need findViewById anymore

File: app/src/main/java/com/lobesoftware/toof/firebase_chat_001/utils/validator/Validator.kt [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        }
+        if (name.length > MAX_LENGTH_NAME_FORMAT) {
+            errorMessages.add(context.resources.getString(R.string.more_than_255_char))
+        }
+        if (name.validWithPattern(ALL_SPECIAL_CHAR_FORMAT)) {
+            errorMessages.add(context.resources.getString(R.string.contain_special_char))
+        }
+        mNotGoodPattern?.let {
+            val isValid = name.validWithPattern(it)
+            val message = if (isValid) "" else context.getString(R.string.contain_invalid_word)
```
Comment (onstonboy) [inline app/src/main/java/com/lobesoftware/toof/firebase_chat_001/utils/validator/Validator.kt]: should be put code in if else in block {}

File: app/src/main/java/com/lobesoftware/toof/firebase_chat_001/utils/validator/Validator.kt [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        for (s in strings) {
+            builder.append(s)
+            builder.append(format)
+        }
+        var result = builder.toString()
+        result = result.substring(0, result.length - format.length)
+        return result
+    }
+
+    companion object {
```
Comment (onstonboy) [inline app/src/main/java/com/lobesoftware/toof/firebase_chat_001/utils/validator/Validator.kt]: put those variable in class Constant -> define all of them to  ez to reuse

File: app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginContract.kt [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package com.lobesoftware.toof.firebase_chat_001.screen.authentication.login
+
+import com.lobesoftware.toof.firebase_chat_001.utils.BasePresenter
+
+interface LoginContract {
+
+    interface View {
+        fun onInputDataValid(errorMessageValidate: ErrorMessageValidate)
```
Comment (NguyenChanhTiep) [inline app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginContract.kt]: onInputDataValid  -> onInputDataInValid

File: app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginPresenter.kt [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+    override fun onStop() {
+    }
+
+    override fun onDestroy() {
+    }
+
+    override fun login(email: String, password: String) {
+        val inputEmailError = mValidator.validateEmail(email)
+        val inputPasswordError = mValidator.validatePassword(password)
```
Comment (NguyenChanhTiep) [inline app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginPresenter.kt]: viết một hàm validate riêng

File: app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginPresenter.kt [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        mView.onInputDataValid(ErrorMessageValidate(inputEmailError, inputPasswordError))
```
Comment (NguyenChanhTiep) [inline app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginPresenter.kt]: Có lỗi mới cần bắn về view, chứ không lỗi không cần bắn về. Vì mình khôgn xử lý gì cả

File: app/src/main/java/com/lobesoftware/toof/firebase_chat_001/utils/validator/Validator.kt [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+            errorMessages.add(context.resources.getString(R.string.contain_special_char))
+        }
+        mNotGoodPattern?.let {
+            val isValid = name.validWithPattern(it)
+            val message = if (isValid) {
+                ""
+            } else {
+                context.getString(R.string.contain_invalid_word)
+            }
+            if (message.isNotBlank()) errorMessages.add(message)
```
Comment (NguyenChanhTiep) [inline app/src/main/java/com/lobesoftware/toof/firebase_chat_001/utils/validator/Validator.kt]: if (isValid) {
errorMessages.add(context.getString(R.string.contain_invalid_word))
}

File: app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginPresenter.kt [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+class LoginPresenter : LoginContract.Presenter {
+
+    private lateinit var mView: LoginContract.View
+    private lateinit var mValidator: Validator
+
+    fun setValidator(validator: Validator) {
+        mValidator = validator
+    }
+
+    override fun setView(v: LoginContract.View) {
```
Comment (daolq3012) [inline app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginPresenter.kt]: v -> view

File: app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginFragment.kt [hunk 12]
```diff
@@ -1,18 +1,74 @@
...
+        (activity?.application as MainApplication).mAppComponent.inject(this@LoginFragment)
+        super.onAttach(context)
+    }
 
     override fun onCreateView(
         inflater: LayoutInflater, container: ViewGroup?,
         savedInstanceState: Bundle?
     ): View? {
-        return inflater.inflate(R.layout.fragment_login, container, false)
+        val view = inflater.inflate(R.layout.fragment_login, container, false)
```
Comment (daolq3012) [inline app/src/main/java/com/lobesoftware/toof/firebase_chat_001/screen/authentication/login/LoginFragment.kt]: mView = inflater.inflate(R.layout.fragment_login, container, false)

**Pull request merged**